evm-adapters/src/sputnik/cheatcodes/debugger.rs

```diff
@@ -0,0 +1,403 @@
+                new_node.idx = self.arena.len();
+                new_node.location = self.arena[entry].children.len();
+                self.arena[entry].children.push(new_node.idx);
+                self.arena.push(new_node);
+            }
+            // we haven't found the parent node, go deeper
+            _ => self.push_node(
+                *self.arena[entry].children.last().expect("Disconnected debug node"),
+                new_node,
+            ),
+        }
+    }
+
+    pub fn flatten(&self, entry: usize, flattened: &mut Vec<(Address, Vec<DebugStep>)>) {
+        let node = &self.arena[entry];
+        flattened.push((node.address, node.steps.clone()));
+        node.children.iter().for_each(|child| {
+            self.flatten(*child, flattened);
+        });
+    }
+}
+
+#[derive(Default, Debug, Clone)]
+/// A node in the arena
+pub struct DebugNode {
+    /// Parent node index in the arena
+    pub parent: Option<usize>,
+    /// Children node indexes in the arena
+    pub children: Vec<usize>,
+    /// Location in parent
+    pub location: usize,
+    /// This node's index in the arena
+    pub idx: usize,
+    /// Address context
+    pub address: Address,
+    /// Depth
+    pub depth: usize,
+    /// The debug steps
+    pub steps: Vec<DebugStep>,
+}
+
+impl DebugNode {
+    pub fn new(address: Address, depth: usize, steps: Vec<DebugStep>) -> Self {
+        Self { address, depth, steps, ..Default::default() }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct DebugStep {
+    pub pc: usize,
+    pub stack: Vec<H256>,
+    pub memory: Memory,
+    pub op: OpCode,
+    pub push_bytes: Option<Vec<u8>>,
+}
+
+impl Default for DebugStep {
+    fn default() -> Self {
+        Self {
+            pc: 0,
+            stack: vec![],
+            memory: Memory::new(0),
+            op: OpCode(Opcode::INVALID, None),
+            push_bytes: None,
+        }
+    }
+}
+
+impl DebugStep {
+    pub fn pretty_opcode(&self) -> String {
+        if let Some(push_bytes) = &self.push_bytes {
+            format!("{}(0x{})", self.op, hex::encode(push_bytes))
+        } else {
+            self.op.to_string()
+        }
+    }
+}
+
+impl Display for DebugStep {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if let Some(push_bytes) = &self.push_bytes {
+            write!(
+                f,
+                "pc: {:?}\nop: {}(0x{})\nstack: {:#?}\nmemory: 0x{}\n\n",
+                self.pc,
+                self.op,
+                hex::encode(push_bytes),
+                self.stack,
+                hex::encode(self.memory.data())
+            )
+        } else {
+            write!(
+                f,
+                "pc: {:?}\nop: {}\nstack: {:#?}\nmemory: 0x{}\n\n",
+                self.pc,
+                self.op,
+                self.stack,
+                hex::encode(self.memory.data())
+            )
+        }
+    }
+}
+
+impl<'b, 'config> ForgeRuntime<'b, 'config> {
```

Comment:
What does this do? Isn't it forwarding everything to the internal runtime? It seems like you could pass the imported `Runtime` to `debug_step` no?

Reply:
junk left over, going to delete